app/src/org/commcare/services/DataSyncCompleteBroadcastReceiver.java

```diff
@@ -0,0 +1,58 @@
+package org.commcare.services;
+
+import static android.app.Activity.RESULT_CANCELED;
+
+import static org.commcare.activities.DispatchActivity.SESSION_REBUILD_REQUEST;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+
+import org.commcare.CommCareApplication;
+import org.commcare.activities.CommCareActivity;
+import org.commcare.activities.DispatchActivity;
+import org.commcare.activities.EntityDetailActivity;
+import org.commcare.activities.EntitySelectActivity;
+import org.commcare.activities.HomeScreenBaseActivity;
+import org.commcare.models.AndroidSessionWrapper;
+import org.commcare.session.CommCareSession;
+import org.commcare.session.SessionDescriptorUtil;
+import org.commcare.session.SessionFrame;
+
+public class DataSyncCompleteBroadcastReceiver extends BroadcastReceiver {
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        AndroidSessionWrapper asw = CommCareApplication.instance().getCurrentSessionWrapper();
+        CommCareSession commcareSession = asw.getSession();
+
+        CommCareActivity activity = (CommCareActivity) context;
+        if (!(activity instanceof HomeScreenBaseActivity)){
+            // We only rebuild the session if the user is not in the Home Screen
+            try {
+                // Rebuild the Session
+                Intent i = new Intent(context, DispatchActivity.class);
+                i.putExtra(SESSION_REBUILD_REQUEST, SessionDescriptorUtil.createSessionDescriptor(commcareSession));
+                if (context instanceof EntityDetailActivity)
+                    addSelectedEntityDatumToSession(activity, commcareSession);
+                context.startActivity(i);
+            } catch (RuntimeException e) {
+                activity.setResult(RESULT_CANCELED);
+            }
+            finally {
+                activity.finish();
+            }
+        }
```

Comment:
I think we can just call - 
````
activity.setResult(HomeScreenBaseActivity.RESULT_RESTART);
activity.finish();
````

here instead of recreating `DispatchActivity` with  This would make the code go back to [onActivityResult for Home Activity](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/HomeScreenBaseActivity.java#L549) which should take care of rebuilding the session by itself.